File: distributed-systems/temporal-workflows/go-code/workflows/workflows.go
package workflows

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

func FoodDeliveryWorkflow(ctx workflow.Context, orderID string) (string, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Starting food delivery workflow", "OrderID", orderID)

	options := workflow.ActivityOptions{
		ScheduleToCloseTimeout: time.Second * 30,
	}
	ctx = workflow.WithActivityOptions(ctx, options)

	var payment, restaurant, driver, delivery string

	// Run activities in order
	if err := workflow.ExecuteActivity(ctx, ChargeCustomer, orderID).Get(ctx, &payment); err != nil {
		return "", fmt.Errorf("payment failed: %w", err)
	}
	if err := workflow.ExecuteActivity(ctx, SendToRestaurant, orderID).Get(ctx, &restaurant); err != nil {
		return "", fmt.Errorf("restaurant failed: %w", err)
	}
	if err := workflow.ExecuteActivity(ctx, AssignDriver, orderID).Get(ctx, &driver); err != nil {
		return "", fmt.Errorf("driver assignment failed: %w", err)
	}
	if err := workflow.ExecuteActivity(ctx, TrackDelivery, orderID).Get(ctx, &delivery); err != nil {
		return "", fmt.Errorf("delivery tracking failed: %w", err)
	}

	return fmt.Sprintf("Order %s -> %s, %s, %s, %s", orderID, payment, restaurant, driver, delivery), nil
}

File: distributed-systems/temporal-workflows/python-code/workflows.py
from temporalio import workflow, activity
import asyncio
from datetime import timedelta


# -------------------------
# Activities (real work)
# -------------------------
@activity.defn(name="charge_customer")
async def charge_customer(order_id: str) -> str:
    print(f"💳 Charging customer for order {order_id}")
    return "payment_confirmed"

@activity.defn(name="send_to_restaurant")
async def send_to_restaurant(order_id: str) -> str:
    print(f"🍴 Sending order {order_id} to restaurant")
    return "restaurant_confirmed"

@activity.defn(name="assign_driver")
async def assign_driver(order_id: str) -> str:
    print(f" Assigning driver for order {order_id}")
    await asyncio.sleep(1)  # simulate waiting
    return "driver_assigned"

@activity.defn(name="track_delivery")
async def track_delivery(order_id: str) -> str:
    print(f"📦 Tracking delivery for order {order_id}")
    return "delivered"

# -------------------------
# Workflow (the recipe)
# -------------------------
@workflow.defn
class FoodDeliveryWorkflow:
    @workflow.run
    async def run(self, order_id: str) -> str:
        payment = await workflow.execute_activity(
            charge_customer, order_id, start_to_close_timeout=timedelta(seconds=10)
        )
        restaurant = await workflow.execute_activity(
            send_to_restaurant, order_id, schedule_to_close_timeout=timedelta(seconds=10)
        )
        driver = await workflow.execute_activity(
            assign_driver, order_id, schedule_to_close_timeout=timedelta(seconds=10)
        )
        delivery = await workflow.execute_activity(
            track_delivery, order_id, schedule_to_close_timeout=timedelta(seconds=10)
        )

        return f"Order {order_id} -> {payment}, {restaurant}, {driver}, {delivery}"
